mkosi/__init__.py

```diff
@@ -602,115 +568,46 @@ def root_partition_name(
-            table += f'size={args.var_size // 512}, type={GPT_VAR}, name="Variable Data Partition"\n'
-            args.var_partno = pn
-            pn += 1
-            run_sfdisk = True
-
-        if args.tmp_size is not None:
-            table += f'size={args.tmp_size // 512}, type={GPT_TMP}, name="Temporary Data Partition"\n'
-            args.tmp_partno = pn
-            pn += 1
-            run_sfdisk = True
-
-    if not is_generated_root(args):
-        table += 'type={}, attrs={}, name="{}"\n'.format(
-            gpt_root_native(args.architecture, args.usr_only).root,
-            "GUID:60" if args.read_only and args.output_format != OutputFormat.gpt_btrfs else "",
-            root_partition_name(args),
-        )
-        run_sfdisk = True
-
-    args.root_partno = pn
-    pn += 1
-
-    if args.verity:
-        args.verity_partno = pn
-        pn += 1
-    else:
-        args.verity_partno = None
-
-    if args.verity == "signed":
-        args.verity_sig_partno = pn
-        pn += 1
-    else:
-        args.verity_sig_partno = None
-
-    return table, run_sfdisk
-
-
-def exec_sfdisk(args: CommandLineArguments, f: BinaryIO) -> None:
-
-    table, run_sfdisk = determine_partition_table(args)
-
-    if run_sfdisk:
-        run(["sfdisk", "--color=never", f.name], input=table.encode("utf-8"))
-        run(["sync"])
-
-    args.ran_sfdisk = run_sfdisk
+def create_partition_table(args: CommandLineArguments) -> PartitionTable:
+    table = PartitionTable.empty(args.gpt_first_lba)
+    no_btrfs = args.output_format != OutputFormat.gpt_btrfs
+
+    for condition, label, size, type_uuid, name, read_only in (
+            (args.bootable and "uefi" in args.boot_protocols,
+             'esp', args.esp_size, GPT_ESP, "ESP System Partition", False),
+            (args.bootable and "bios" in args.boot_protocols,
+             'bios', BIOS_PARTITION_SIZE, GPT_BIOS, "BIOS Boot Partition", False),
+            (args.xbootldr_size is not None,
+             'xbootldr', args.xbootldr_size, GPT_XBOOTLDR, "Boot Loader Partition", False),
+            (args.swap_size is not None,
+             'swap', args.swap_size, GPT_SWAP, "Swap Partition", False),
+            (no_btrfs and args.home_size is not None,
+             'home', args.home_size, GPT_HOME, "Home Partition", False),
+            (no_btrfs and args.srv_size is not None,
+             'srv', args.srv_size, GPT_SRV, "Server Data Partition", False),
+            (no_btrfs and args.var_size is not None,
+             'var', args.var_size, GPT_VAR, "Variable Data Partition", False),
+            (no_btrfs and args.tmp_size is not None,
+             'tmp', args.tmp_size, GPT_TMP, "Temporary Data Partition", False),
+            (not is_generated_root(args),
+             'root', args.root_size,
+             gpt_root_native(args.architecture, args.usr_only).root,
+             root_partition_name(args),
+             args.read_only)):
+
+        if condition and size is not None:
```

Comment:
this means you check the size agaist None twice each time? via the list above and here again?

I presume the first field of the list is resolved at the moment the list is put together? i.e. greedely? by moving this into lambda funcs we could resolve it lazily? would that be nicer?

Reply:
I think this would make the code more complicated for no gain. We go through the whole list anyway, so evaluating a few trivial conditional eagerly is probably faster then going lazy evaluation later on.